db/migrate/20190213082646_add_approval_to_listings.rb

```diff
@@ -0,0 +1,6 @@
+class AddApprovalToListings < ActiveRecord::Migration[5.1]
+  def change
+    add_column :listings, :state, :string, default: 'approved'
+    add_index :listings, :state
```

Comment:
This index alone doesn't help much. Listings are normally selected first by community_id, then by other columns. The main listings query that must be optimised is the one on the homepage, happening here: https://github.com/sharetribe/sharetribe/blob/ability-to-have-pre-approved-listings/app/services/listing_index_service/search/database_search_helper.rb#L29

That query translates to:

```sql
SELECT  `listings`.* FROM `listings` WHERE `listings`.`community_id` = 7198 AND `listings`.`deleted` = 0 AND `listings`.`open` = 1 AND `listings`.`state` = 'approved' AND (valid_until IS NULL OR valid_until > '2019-02-25 15:22:38.617921') ORDER BY listings.sort_date DESC;
```

The existing index that covers this is: `homepage_query_valid_util` on `(community_id, open, valid_until, sort_date, deleted))`. So now we need a new one that contains also `state`. The state needs to be before the `sort_date` column, as it is used for final ordering. I suggest the following index:

```sql
CREATE INDEX listings_homepage_query ON listings(community_id, open, state, deleted, valid_until, listing_shape_id, sort_date);
```

Column order in the index should be slightly better and I suggest adding the `listing_shape_id`, which sometimes is used in the homepage query.

The new index can be added and the old `homepage_query` and `homepage_query_valid_until` can be removed, as this index should cover all uses of the old ones. Make sure the migration first adds the new index and then deletes the old ones, so that the database is not left without proper index at any point.


Similar situation with the old `updates_email_at` index. It should cover the `state` column. Add new index:

```sql
CREATE INDEX updates_email ON listings(community_id, open, state, deleted, valid_until, updates_email_at);
```

While checking that, I noticed that `deleted` listings don't seem to be excluded from update email. @aivils should we add check for `deleted` in the `currently_open` scope? Is there any place that should still be able to see deleted listings? It seems to be easy to forget including `.exist` in all places where listing scope is used. Or not?

Reply:
The speedup for the homepage query with the index I suggest is ~9-11% in some non-scientific test I did.